gridstatus/ieso.py

```diff
@@ -0,0 +1,499 @@
+        # Initialize a list to store the parsed data
+        data = []
+
+        # Parse the XML file
+        root = ET.fromstring(r.content)
+
+        published_time = root.find(".//CreatedAt", NAMESPACES_FOR_XML).text
+
+        # Extracting data for each ZonalDemands within the Document
+        for zonal_demands in root.findall(".//ZonalDemands", NAMESPACES_FOR_XML):
+            delivery_date = zonal_demands.find(
+                ".//DeliveryDate",
+                NAMESPACES_FOR_XML,
+            ).text
+
+            for zonal_demand in zonal_demands.findall(
+                ".//ZonalDemand/*",
+                NAMESPACES_FOR_XML,
+            ):
+                # The zone name is the tag name without the namespace
+                zone_name = zonal_demand.tag[(zonal_demand.tag.rfind("}") + 1) :]
+
+                for demand in zonal_demand.findall(".//Demand", NAMESPACES_FOR_XML):
+                    hour = demand.find(".//DeliveryHour", NAMESPACES_FOR_XML).text
+                    energy_mw = demand.find(".//EnergyMW", NAMESPACES_FOR_XML).text
+
+                    data.append(
+                        {
+                            "DeliveryDate": delivery_date,
+                            "Zone": zone_name,
+                            "DeliveryHour": hour,
+                            "EnergyMW": energy_mw,
+                        },
+                    )
+
+        df = pd.DataFrame(data)
+
+        # Convert columns to appropriate data types
+        df["DeliveryHour"] = df["DeliveryHour"].astype(int)
+        df["EnergyMW"] = df["EnergyMW"].astype(float)
+        df["DeliveryDate"] = pd.to_datetime(df["DeliveryDate"])
+
+        df["Interval Start"] = (
+            # Need to subtract 1 from the DeliveryHour since that represents the
+            # ending hour of the interval. (1 represents 00:00 - 01:00)
+            df["DeliveryDate"]
+            + pd.to_timedelta(df["DeliveryHour"] - 1, unit="h")
+        ).dt.tz_localize(self.default_timezone)
+
+        df["Interval End"] = df["Interval Start"] + pd.Timedelta(hours=HOUR_INTERVAL)
+
+        # Pivot the table to wide
+        pivot_df = df.pivot_table(
+            index=["Interval Start", "Interval End"],
+            columns="Zone",
+            values="EnergyMW",
+            aggfunc="first",
+        ).reset_index()
+
+        pivot_df["Publish Time"] = pd.Timestamp(
+            published_time,
+            tz=self.default_timezone,
+        )
+
+        pivot_df = utils.move_cols_to_front(
+            pivot_df,
+            [
+                "Interval Start",
+                "Interval End",
+                "Publish Time",
+                "Ontario",
+            ],
+        )
+
+        pivot_df.columns.name = None
+
+        col_mapper = {
+            col: f"{col} Load Forecast" for col in ["Ontario", "East", "West"]
+        }
+
+        pivot_df = pivot_df.rename(columns=col_mapper)
+
+        # Return all the values from the start of today onwards
```

Comment:
when we have a data source that has a publish time, consider the input date to used to filter against the publish time. so, no need to filter the intervals.

the behavior of filter against intervals gets confusing because technically there are multiple publish times for each of these intervals.

Reply:
Makes sense. I've updated the filtering.